### PR TITLE
Fix model plot display when factor is set for grouped PHA data (issue #1779)

### LINKED ISSUE #1779: plot_model fails when factor != 0 and data is filtered
```
sherpa-4.15.0> load_pha("3c273.pi")
WARNING: systematic errors were not found in file '3c273.pi'
statistical errors were found in file '3c273.pi' 
but not used; to use them, re-read with use_errors=True
read ARF file 3c273.arf
read RMF file 3c273.rmf
WARNING: systematic errors were not found in file '3c273_bg.pi'
statistical errors were found in file '3c273_bg.pi' 
but not used; to use them, re-read with use_errors=True
read background file 3c273_bg.pi

sherpa-4.15.0> set_source(xsphabs.gal * powlaw1d.pl)

sherpa-4.15.0> set_analysis(1, "energy", factor=1)

sherpa-4.15.0> plot_model()
```

sherpa1

If I now filter the data and try to plot the model it fails with a mis-match.

```
sherpa-4.15.0> notice(0.5, 6)
dataset 1: 0.00146:14.9504 -> 0.4672:6.57 Energy (keV)

sherpa-4.15.0> import sys; sys.tracebacklimit = 10

sherpa-4.15.0> plot_model()
  File "/home/dburke/miniconda3/envs/ciao415-py310/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3433, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-8-985f7c7e70f6>", line 1, in <module>
    plot_model()
  File "<string>", line 1, in plot_model
  File "/home/dburke/miniconda3/envs/ciao415-py310/lib/python3.10/site-packages/sherpa/ui/utils.py", line 13239, in plot_model
    plotobj = self.get_model_plot(id, recalc=not replot)
  File "/home/dburke/miniconda3/envs/ciao415-py310/lib/python3.10/site-packages/sherpa/astro/ui/utils.py", line 10452, in get_model_plot
    plotobj.prepare(data, self.get_model(id), self.get_stat())
  File "/home/dburke/miniconda3/envs/ciao415-py310/lib/python3.10/site-packages/sherpa/astro/plot.py", line 261, in prepare
    data.mask = old_mask
  File "/home/dburke/miniconda3/envs/ciao415-py310/lib/python3.10/site-packages/sherpa/utils/__init__.py", line 171, in __setattr__
    object.__setattr__(self, name, val)
  File "/home/dburke/miniconda3/envs/ciao415-py310/lib/python3.10/site-packages/sherpa/astro/data.py", line 1871, in mask
    nexp = len(self.get_y(filter=False))
  File "/home/dburke/miniconda3/envs/ciao415-py310/lib/python3.10/site-packages/sherpa/astro/data.py", line 4150, in get_y
    val = self._fix_y_units(val, filter, response_id)
  File "/home/dburke/miniconda3/envs/ciao415-py310/lib/python3.10/site-packages/sherpa/astro/data.py", line 4132, in _fix_y_units
    val *= scale
ValueError: operands could not be broadcast together with shapes (46,) (41,) (46,) 

sherpa-4.15.0> 
```

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Ensure that model plots can be created for PHA data when set_analysis routine is used to set factor  > 0 and the data is grouped. Fixes #1779 and #1784

# Details

The factor setting comes in to the very-end of the internal "fix units" routine for DataPHA objects, and obviously missed-out on the recent work to improve filtering behavior (eg #1477) as we didn't have a test that triggered the problem. So a test has been added to try and cover the problem (at a  lower level than found in #1779, which used the UI layer). Note that we want to check

- grouped and ungrouped data
- filtered and unfiltered data
- plot_model and  plot_fit (the model part) behavior, which is handled by the two `sherpa.astro.plot` model classes

In working on this I came across #1784 so i have addressed this too. As this is all for "approximate" data it may not matter exactly how we calculate the mid-point of each group, but I prefer the newer version.

We should review the DataPHA class for data access to channel/group values (e.g. `get_x` vs `_get_ebins` vs ...) but that is a significant rework but I'm still birnt out by #1477 and related work, so I do not plan on this for quite some time.

I think this should be looked at and then #1828

### CODE REVIEW

**Comment by DougBurke** [comment]
@anetasie I've assigned this to you to try out before I give it to @hamogu for a code review.

**Comment by DougBurke** [comment]
I've assigned this to @hamogu so I don't forget to do it later (or am out of office) but it needs @anetasie review first.

**Review by anetasie — APPROVED** [review]
It looks good! Thanks

File: sherpa/astro/data.py [hunk 1]
```diff
@@ -4192,47 +4192,59 @@ def _fix_y_units(self, val, filter=False, response_id=None):
...
-                ebin = hc / elo - hc / ehi
-            elif self.units == 'channel':
-                ebin = ehi - elo
+            return xlo, xhi
+
+        if self.grouped or self.rate:
+            xlo, xhi = get_bin_edges()
+            if self.units == 'wavelength':
+                dx = hc / xlo - hc / xhi
             else:
```
Comment (hamogu) [inline sherpa/astro/data.py:4219]: Suggest added comment for future readers
```suggestion
            else:  # Could be "energy" or "channel"
```

**Review by hamogu — APPROVED** [review]
Just one tiny suggestion to add a doc for future readers of this code. (But optional, so approving already.)

Otherwise, the code looks sensible. If this is what we want to do, it should work. And Aneta has already signed off that this is what we want to do, so fine with me.

**Pull request merged**